server/src/test/python/cromwell-az-e2e-test/cromwell_az_e2e_test/az_e2e.py

```diff
@@ -0,0 +1,229 @@
+    uri = f"{leo_url}/api/apps/v2/{workspaceId}?includeDeleted=false"
+
+    headers = {"Authorization": f"Bearer {bearer_token}",
+               "accept": "application/json"}
+
+    response = requests.get(uri, headers=headers)
+    status_code = response.status_code
+    handle_failed_request(response, f"{status_code} - Error retrieving details for {workspaceId}", 200)
+    output_message(f"Successfully retrieved details.\n{response.text}")
+    response = response.json()
+
+    app_url = ""
+    app_type = "CROMWELL" if app != 'wds' else app.upper()
+    output_message(f"App type: {app_type}")
+    for entries in response: 
+        if entries['appType'] == app_type and entries['proxyUrls'][app] is not None:
+            if(entries['status'] == "PROVISIONING"):
+                output_message(f"{app} is still provisioning")
+                break
+            output_message(f"App status: {entries['status']}")
+            app_url = entries['proxyUrls'][app]
+            break 
+
+    if app_url is None: 
+        output_message(f"{app} is missing in current workspace")
+    else:
+        output_message(f"{app} url: {app_url}")
+
+    return app_url
+
+def submit_workflow_to_cromwell(app_url, workflow_test_name):
+    absolute_file_path = os.path.dirname(__file__)
+    workflow_source_path = os.path.join(absolute_file_path, './workflow_files/hello.wdl')
+    workflow_inputs_path = os.path.join(absolute_file_path, './workflow_files/hello.inputs')
+    workflow_endpoint = f'{app_url}/api/workflows/v1'
+    headers = {"Authorization": f'Bearer {bearer_token}',
+              "accept": "application/json",
+    }
+    with open(workflow_source_path, 'rb') as hello_wdl:
+        with open(workflow_inputs_path, 'rb') as hello_inputs:
+            files = {
+                'workflowSource': ('hello.wdl', hello_wdl, 'application/octet-stream'),
+                'workflowInputs': ('hello.inputs', hello_inputs, 'application/octet-stream'),
+                'workflowType': 'WDL',
+                'workflowTypeVersion': '1.0'
+            }
+            response = requests.post(workflow_endpoint, headers=headers, files=files)
+            handle_failed_request(response, f"Error submitting workflow to Cromwell for {workflow_test_name}", 201)
+            output_message(response.json())
+            return response.json()
+
+def get_workflow_information(app_url, workflow_id):
+    workflow_endpoint = f'{app_url}/api/workflows/v1/{workflow_id}/status'
+    headers = {"Authorization": f'Bearer {bearer_token}',
+              "accept": "application/json"}
+    output_message(f"Fetching workflow status for {workflow_id}")
+    response = requests.get(workflow_endpoint, headers=headers)
+    handle_failed_request(response, f"Error fetching workflow metadata for {workflow_id}")
+    output_message(response.json())
+    return response.json()
+
+def get_completed_workflow(app_url, workflow_ids, max_retries=4, sleep_timer=60 * 2):
```

Comment:
How do you feel about letting this run forever, and having the test fail when the token eventually expires? We could detect that and print an error that makes it obvious what's happening. 

Partially depends on the outcome of tomorrow's conversation, of course - if it's important that this test check workspace deletion, we'll need to try harder to make sure we attempt it (though IMO if we need to test workspace deletion we should do it in a different test).

Reply:
I'm fine with that, I didn't realize until the last couple of workflow that the deletion steps don't have a consistent runtime. Sometimes they finish in 5 minutes, but other times they take 10+ min, so terminating on a 401 makes sense.